native/libcst/src/parser/grammar.rs

```diff
@@ -0,0 +1,2930 @@
+        rule args() -> Vec<Arg<'a>>
+            = first:_posarg()
+                rest:(c:comma() a:_posarg() {(c, a)})*
+                kw:(c:comma() k:kwargs() {(c, k)})? {
+                    let (trail, kw) = kw.map(|(x,y)| (Some(x), Some(y))).unwrap_or((None, None));
+                    concat(
+                        comma_separate(first, rest, trail),
+                        kw.unwrap_or_default(),
+                    )
+            }
+            / kwargs()
+
+        rule _posarg() -> Arg<'a>
+            = a:(starred_expression() / e:named_expression() { make_arg(e) })
+                !lit("=") { a }
+
+        rule starred_expression() -> Arg<'a>
+            = star:lit("*") e:expression() { make_star_arg(star, e) }
+
+        rule kwargs() -> Vec<Arg<'a>>
+            = sitems:separated(<kwarg_or_starred()>, <comma()>)
+                scomma:comma()
+                ditems:separated(<kwarg_or_double_starred()>, <comma()>) {
+                    concat(
+                        comma_separate(sitems.0, sitems.1, Some(scomma)),
+                        comma_separate(ditems.0, ditems.1, None),
+                    )
+            }
+            / items:separated(<kwarg_or_starred()>, <comma()>) {
+                    comma_separate(items.0, items.1, None)
+            }
+            / items:separated(<kwarg_or_double_starred()>, <comma()>) {
+                    comma_separate(items.0, items.1, None)
+            }
+
+        rule kwarg_or_starred() -> Arg<'a>
+            = _kwarg()
+            / starred_expression()
+
+        rule kwarg_or_double_starred() -> Arg<'a>
+            = _kwarg()
+            / star:lit("**") e:expression() { make_star_arg(star, e) }
+
+        rule _kwarg() -> Arg<'a>
+            = n:name() eq:lit("=") v:expression() {
+                make_kwarg(n, eq, v)
+            }
+
+        rule atom() -> Expression<'a>
+            = n:name() { Expression::Name(n) }
+            / n:lit("True") { Expression::Name(make_name(n)) }
+            / n:lit("False") { Expression::Name(make_name(n)) }
+            / n:lit("None") { Expression::Name(make_name(n)) }
+            / &(tok(STRING, "") / tok(FStringStart, "")) s:strings() {s.into()}
+            / n:tok(Number, "NUMBER") { make_number(n) }
+            / &lit("(") e:(tuple() / group() / (g:genexp() {Expression::GeneratorExp(g)})) {e}
+            / &lit("[") e:(list() / listcomp()) {e}
+            / &lit("{") e:(dict() / set() / dictcomp() / setcomp()) {e}
+            / lit("...") { Expression::Ellipsis(Ellipsis {lpar: vec![], rpar: vec![]})}
+
+        rule strings() -> String<'a>
+            = s:(str:tok(STRING, "STRING") t:&_ {(make_string(str), t)}
+                / str:fstring() t:&_ {(String::Formatted(str), t)})+ {
+                make_strings(s)
+            }
+
+        rule tuple() -> Expression<'a>
+            = lpar:lpar() first:star_named_expression() &lit(",")
+                rest:(c:comma() e:star_named_expression() {(c, e)})*
```

Comment:
[nit / curiosity] The official grammar uses `star_named_expression ', ' [star_named_expressions]` here.

This code is needed to handle empty tuples... I'm actually having trouble figuring out where the official grammar does that.

Reply:
I remember being confused about this, but I don't remember if I found how this works in CPython. Maybe I just gave up and implemented it **MY WAY** :stuck_out_tongue_closed_eyes: